### ISSUE TITLE
User-Agent Client Hints in v89

### ISSUE THREAD

**Issue by wchen342** [bromite/bromite#973]
### Is your feature request related to privacy?

I think so, from the description it leaks a lot of information about the device.

### Is there a patch available for this feature somewhere?

No.

### Describe the solution you would like

Remove the functionality or add a dummy over all devices.

### Describe alternatives you have considered

None.

I came across this at Eloston/ungoogled-chromium#1422. It is only available on chromium-based browsers for now, and from the linked article it seems extensions will not be able to prevent this (we don't have extensions anyway..). Removing this may break websites but the privacy benefits shall overweigh that I think.

cc @uazo

**Comment by csagan5** [comment]
This feature was originally developed to stop client-side using browser version strings; did it change into something else now?

I am in favor of reporting capabilities, not in favor of reporting anything which is privacy-breaking.

Capabilities-reporting can be used for profiling as explained in those articles; in general I would say: if client-side can detect the capability by other means there is no need to hide/obfuscate the hint.

**Comment by bat999** [comment]
Hi

"...from the description it leaks a lot of information about the device"

I've been using User-Agent Client Hints for a while now.
It's hidden behind flag "Experimental Web Platform features".

This website shows how much info is given out with UA-CH enabled...
https://browserleaks.com/client-hints

If UA-CH is automatically enabled in v89 then maybe a flag could be considered with option to disable it.
(But only if it really is considered to be a privacy hazard).
"Enable User Agent Client Hints"

**Comment by wchen342** [comment]
This is what I got on my v88 desktop uc. It shows 4g because I'm using my phone's network:

![1](https://user-images.githubusercontent.com/6406743/110120252-b5a4e500-7db4-11eb-8dd8-6da657c392f7.png)

So basically at least on uc it is not really working for the lower half. However it is the upper half I'm more concerned, including the network type and speed. The memory detection, according to the documentation, is very coarse but still of concern.

**Comment by wchen342** [comment]
I read the IETD draft for client hints. It doesn't specify what information can be used as hints and what cannot, only a vague "security concern" section: [https://httpwg.org/http-extensions/ioggstream-1392/draft-ietf-httpbis-client-hints.html](https://httpwg.org/http-extensions/ioggstream-1392/draft-ietf-httpbis-client-hints.html).

Unless I misunderstood something, this means is it possible for the vendors to add arbitrary fields on their own, providing they have enough influence and websites will follow.

**Comment by csagan5** [comment]
I already explained what the work to be done is here:
* find out which of these hints can be determined without the explicit hint

For all those which can be detected by other means, you have to fix both or neither. Hiding the client hint will just give a false sense of privacy.

This is the same reason why hiding the WebGL vendor is moot: vendor can be determined in other ways.

**Comment by wchen342** [comment]
I will close this for now until we get a clearer view on how this feature is actually used in the wild.

**Comment by bat999** [comment]
> ... I think these are only the demo sites' choices of parameters...


Yes, I understand.

But what if the server were to request some weird sinister information...
Are you Democrat or are you Republican?
Are you Christian or are you Muslim?

Surely it would need to send requests like this...
Accept-CH: Sec-CH-UA-Full-foo
or
Accept-CH: Sec-CH-UA-Full-bar
or
Accept-CH: Sec-CH-UA-Full-foo, Sec-CH-UA-bar

I suppose chromium-based browsers will be coded so they only answer requests from the spec.
So far, these seem to be...
Sec-CH-UA
Sec-CH-UA-Full-Version
Sec-CH-UA-Platform
Sec-CH-UA-Platform-Version
Sec-CH-UA-Arch
Sec-CH-UA-Model
Sec-CH-UA-Mobile

I don't know how they would react to Accept-CH: Sec-CH-UA-Full-foo, Sec-CH-UA-bar

Maybe they will refuse connection and give an error report.

Or maybe they will persevere and show null strings.
Sec-CH-UA-Full-foo ""
Sec-CH-UA-Full-bar ""

(I do know that Chromium mobile shows null string for Sec-CH-UA-Arch)
(Look here --> https://github.com/WICG/ua-client-hints/issues/205)

Also I found this website...
https://manytools.org/http-html-text/http-request-headers/
For me it shows...
Sec-Ch-Ua, Sec-Ch-Ua-Mobile
Because #enable-experimental-web-platform-features is Enabled
and
Save-Data : on
Because #enable-save-data-header is Enabled.
:-)

**Comment by PF4Public** [comment]
@csagan5 @wchen342 Have you had any further discussions on the topic? Had anything changed since then?

**Comment by wchen342** [comment]
Not really. Although there were reports that wrong hints break certain websites. I think csagan5's conclusion still applies.

**Comment by csagan5** [comment]
Related issues/PRs:
* #2465
* #1830
* #973

In #1830 I accepted to more radically disable them and they have been like that since then; there have been no reports of breakage related to this but I also think we should stick to the [original position](https://github.com/bromite/bromite/issues/973#issuecomment-791433008); @uazo is it time to revisit the premises of #1830? (see description there)
> [client hints] would make it possible to make the purposes of the ua reduction ineffective.
in particular, the management of accept ch in the always incognito mode, which by keeping the requests of the sites in the content settings, would allow to know if a user has already visited the site in the same session.
currently the management of client hints is under development and therefore it could be possible that the patch will need to be revised over time.

Basically client hints that are available via JavaScript should be allowed, provided that there is no privacy degradation as originally described above.

**Comment by uazo** [comment]
Please open a new issue

**Comment by PF4Public** [comment]
Thanks everyone for getting back.

@csagan5 Thanks for the especially detailed reply!

So to sum everything up, bromite did eventually merge a PR to disable those hints, but at present the idea is to revert those changes back per @csagan5 's reasoning above. Did I get it correct?

@uazo Who was your request directed at?

**Comment by uazo** [comment]
> Who was your request directed at?

to @csagan5 if can please open a new issue ...

> Basically client hints that are available via JavaScript should be allowed

... to discuss this, I would remove support for `navigator.userAgentData` https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData
see also https://wicg.github.io/ua-client-hints/#fingerprinting

>  the management of accept ch in the always incognito mode, which by keeping the requests of the sites in the content settings, would allow to know if a user has already visited the site in the same session.

... and for this: testing with https://browserleaks.com/client-hints everything seems all OK to me

**Comment by csagan5** [comment]
> So to sum everything up, bromite did eventually merge a PR to disable those hints, but at present the idea is to revert those changes back per @csagan5 's reasoning above. Did I get it correct?

Some clarity needs to be done first; the goal is to have the same amount of information exposed via UA, client hint headers and Javascript, and to have that information be the absolute minimum. We also need to check if the "client hint not supported" exception mechanism was implemented upstream by now.

> Please open a new issue

The issue to be fixed is not yet clear to me, that is why I asked "is it time to revisit the premises of https://github.com/bromite/bromite/pull/1830? (see description there)"

I would open one only once there is a clear problem to solve.

> I would remove support for `navigator.userAgentData` https://developer.mozilla.org/en-US/docs/Web/API/NavigatorUAData
> see also https://wicg.github.io/ua-client-hints/#fingerprinting

This object, together with client hints and the "old" user agent header, should all contain the same information; why remove it if we can make it contain the bare minimum?

Example from desktop posing as mobile:
```json
{
  "brands":[
     {"brand":"Not?A_Brand","version":"8"},{"brand":"Chromium","version":"108"},{"brand":"Google Chrome","version":"108"}
   ],
   "mobile":true,
   "platform":"Android"}
```

**Comment by uazo** [comment]
> This object, together with client hints and the "old" user agent header, should all contain the same information; why remove it if we can make it contain the bare minimum?

Specs does not speak of an obligation for user agents to have clienhints.
in bromite has half the implementation. my proposal is to remove it completely, including any http headers.
are you clear on the objectives of the client-hints? because maybe that's the thing I'm missing.

**Comment by uazo** [comment]
@PF4Public perhaps I have not explained myself well.

what i'm saying is that other browsers also don't support (now) client-hints (safari, firefox and brave itself), why should bromite considering add it now? (to me) it looks like a way to reduce the information from the user agent (required by the specifications) and move it to the clienthints. 
what do you think they are for? for me it is for switch from a passive to an active fingerprint to potentially block it, which we should do immediately.

**Comment by PF4Public** [comment]
> (to me) it looks like a way to reduce the information from the user agent (required by the specifications) and move it to the clienthints.
> what do you think they are for? for me it is for switch from a passive to an active fingerprint to potentially block it, which we should do immediately.

That corresponds exactly to my vision of client-hints! But like @csagan5 had put it, simply removing them would still leave some of the information available via JS, wouldn't it be so?

Since you have merged https://github.com/bromite/bromite/pull/1830 and there seems to be no intent on reverting that, I hope, it would be possible to use your patch with ungoogled-chromium after we improve our license notices and delineate better ungoogled-chromium's BSD code and Bromite's GPL.

**Comment by uazo** [comment]
> simply removing them would still leave some of the information available via JS, wouldn't it be so?

the patch is not complete, I'm sorry, but I am learning and slowly understanding what is the best approach to development, which includes reading (and understanding) the specifications, which I probably had not done.
in this specific case, the removal of the javascript part and the network is missing in order to actually have the client hints disabled.

> I hope, it would be possible to use your patch with ungoogled-chromium

of course, use it! and the fact that you are asking is nice but not mandatory!
the only thing is that you have to figure out how to put GPL code into UC, to respect my desire to keep the code open, although it seems to be enough that UC and any fork derived from UC, if it uses that patch, must give the user the possibility to recompile it (which seems to me already possible :).
